### ISSUE TITLE
CopiedFailure.getTraceback should include the result of CopiedFailure.getErrorMessage even if unsafeTracebacks is not set

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4600]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| titty reported|
|-|-|
|Trac ID|trac#581|
|Type|enhancement|
|Created|2004-04-03 22:55:31Z|
|Branch|https://github.com/twisted/twisted/tree/copiedfailure-printtraceback-581|

Attachments:

* [581-pb.CopyableFailure.getStateToCopy.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9f6/9f682df245668969bbcd5395bdc2882591eeecde/9f67b8cf81571dafff0842326ff9542187c3c46c.patch) (564 bytes) - added by mattcan on 2011-04-14 03:04:55Z - 
* [581.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9f6/9f682df245668969bbcd5395bdc2882591eeecde/622d87150a46e6b5eb393cd9cb9b3042a74d6266.patch) (1743 bytes) - added by retenodus on 2011-07-29 13:29:30Z - Replacement of Traceback Unavailable by the value of error, modification of tests

<details><summary>Searchable metadata</summary>

```
trac-id__581 581
type__enhancement enhancement
reporter__titty titty
priority__normal normal
milestone__ 
branch__branches_copiedfailure_printtraceback_581 branches/copiedfailure-printtraceback-581
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__easy easy
time__1081032931000000 1081032931000000
changetime__1478740488452472 1478740488452472
version__ 
owner__exarkun exarkun
cc__warner cc__titty cc__mattcan cc__thijs cc__retenodus cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
When an error occurs using pb, one sees messages like the following:
CopiedFailure: [Failure instance: Traceback from remote host -- Traceback
unavailable

They are not very informative, more information is available and should also be
printed:


Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py   (revision 10463)
+++ twisted/python/failure.py   (working copy)
@@ -250,7 +250,7 @@
         return "&lt;%s %s>" % (self.__class__, self.type)
 
     def __str__(self):
-        return "[Failure instance: %s]" % self.getBriefTraceback()
+        return "[Failure instance: %s: %s: %s]" % (self.type,
self.getErrorMessage(), self.getBriefTraceback())
 
     def __getstate__(self):
         """Avoid pickling objects in the traceback.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Assign to warner
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
Could you look at broker.unsafeTracebacks and see if it provides what you want?
The idea is that you don't want to reveal a whole lot about the remote end by
default. If you do, you set .unsafeTracebacks=True .

That said, I'm not sure it's implemented correctly. (I've been trying to improve
on this in newpb, where it is easier to control how Failures get serialized).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
Never seen that unsafeTracebacks flag before. Thanks for the tip.
But the problem occurs when unsafeTracebacks is not set. I'd like to print a
failure object, and all I see is "[Failure instance: Traceback from remote host
-- Traceback unavailable]". I could print failure.type,
failure.getErrorMessage()..but it's just not as simple as it ought to be.

btw, is newpb already usable and what other advantages does it provide?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
when one doesn't handle errors in deferreds (using pb, i.e. CopiedFailure),
twisted from svn (r13537) just prints:

'Unhandled error in Deferred:'

no traceback, no getErrorMessage(), nothing else.
guess this isn't helpful at all...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Regardless of whether `unsafeTracebacks` is set, the copied failure has the exception error message.  There's no reason to exclude it from the result of `getTraceback` if it's available.

This should be a simple matter of adjusting the string created by `CopiedFailure.printTraceback`, or possibly by changing what `CopyableFailure.getStateToCopy` puts into the 'traceback' key of the state dictionary when `unsafeTracebacks` is not set.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="squiddo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|squiddo commented|
|-|-|

Apologies for the same mistake - using python 2.5 for working on
twisted now.

Added deprecation warnings suppression for the Zipstream tests for
unzip and unzipIter.

Please review.

**Comment by twisted-trac** [comment]
|<img alt="squiddo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|squiddo commented|
|-|-|

Ignore last message - wrong ticket.

**Comment by twisted-trac** [comment]
|<img alt="mattcan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|mattcan commented|
|-|-|

Dropped the check to see if unsafeTracebacks was set.

Existing unit tests says the function is deprecated so I didn't create a unit test. I can go back and create a test if need be.

```
    def test_printTraceback(self):
        """
        L{pb.printTraceback} is deprecated.
        """
        self.callDeprecated(
            Version("twisted", 8, 2, 0),
            pb.printTraceback,
            "printTraceback deprecation fake traceback value")

```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to mattcan|
|-|-|

Thanks for working on this mattcan.  I think this patch makes too big a change, though.  It entirely disables the `unsafeTracebacks` feature, so complete tracebacks are always sent to the peer.  We don't want to do that (or if we did, we could delete a whole lot more code than your patch deleted ;).  Instead, we want to preserve the stringified exception message even if `unsafeTracebacks` is set to `False`.

This could be done by sending a fake traceback which includes no stack frames but does include the stringified exception message.  This would be a change to `CopyableFailure.getStateToCopy`.

Or it could be done by changing `CopiedFailure.printTraceback` to _also_ print out `self.value` instead of only printing `self.traceback`.

I would prefer the former, I think, since it would be nice to see `CopiedFailure` vanish someday.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to retenodus|
|-|-|

Thanks!  This seems closer, but it's wrong for _traceback_ to be associated with the string representation of the exception value.  Also, the exception value is already available as a string associated with the _value_ key.

The only thing that needs to change is for the _reporting_ of the value to be improved.  The base `Failure` type will write out `self.value`, it's only `CopiedFailure` that doesn't do this in its override of the `printTraceback` method.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [32679]) Branching to 'copiedfailure-printtraceback-581'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [32680]) Write out the exception type and value in CopiedFailure.printTraceback

refs [#4600](https://github.com/twisted/twisted/issues/4600)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/copiedfailure-printtraceback-581)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_19"></a>|@jesstess set owner to @exarkun|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [32811]) Merge copiedfailure-printtraceback-581

Author: exarkun
Reviewer: jesstess
Fixes: [#4600](https://github.com/twisted/twisted/issues/4600)

Enhance `twisted.spread.pb.CopiedFailure`, the object used to represent `Failure`
instances sent over PB, so that when it prints itself, it prints out the type
and value of the wrapped exception.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_21"></a>|@rodrigc commented|
|-|-|

In changeset 0797bdc094947e487db8cb7b1c2a87e001271d35
```
#!CommitTicketReference repository="" revision="0797bdc094947e487db8cb7b1c2a87e001271d35"
Merge pull request #581 from twisted/8888-rodrigc-dirdbm

Author: rodrigc
Reviewer: moshez wsanchez

Port dirdbm to Python 3
```
